### ISSUE TITLE
Q: Creating a NgModule for aot?

### ISSUE THREAD

**Issue by m-abs** [manekinekko/angular2-aot-demo#10]
Thank you very much for this demo, it helps a lot.

However I wonder how to best write an NgModule to be distributed via npmjs that can be used with an aot-build like yours?

I started out having a NgModule that I created and build with gulp-typescript, this couldn't be imported by your demo-project, I got this error:
```bash
Error: Unexpected value 'MyModule' imported by the module 'AppModule'
```

I quickly found that it was because of missing metadata generated by `ngc`, so I ended up including  `ngc` in our project's gulp-config and spend alot of time with trail-and-error to get it to build.
Result can be seen here: https://github.com/m-abs/typescript-baseproject/tree/ng2-ngc

But I don't like my solution all that much, as I run the `ngc`-command with `child_process.exec` from my gulpfile.js and runs the output through `tsc` afterwards.

I've been trying to find a guide or something that could explain what a NgModule npm-package should contain and how it should be build, but all I can find it various demos and seed projects for building a full app, not a component.

**Comment by manekinekko** [comment]
Hi @m-abs 
A short answer would be: your NgModule doesn't need to do anything special. Actually, it can't. AOT happens on build step. And webpack runs Tree Shaking on the generated *.ts file.
BTW, in your setup, you don't benefit from Tree Shaking, since you're transpiling using`tsc`.

Also, I think that using `child_process.exec` to run `ngc` is your only solution with `gulp`. I don't know how you can do it differently.

Anyways, building and distributing an ng2 NgModule is no different from sharing a typescript package.

I Hope that makes sense.

**Comment by m-abs** [comment]
Hi @manekinekko 

Thanks you for taking time to answer :+1: 

> A short answer would be: your NgModule doesn't need to do anything special.

When my module is build and distributed without the metadata generated by `ngc` I get this error when installed in your demo project:
```bash
Error: Unexpected value 'MyModule' imported by the module 'AppModule'
```
This is the reason I tried to do as I did.

I've tried writting an example to show what I'm experience more clearly.
[with ngc](https://github.com/m-abs/angular2-aot-demo/tree/with-ngc)
I can build the demo project just as expected and my module is loaded. The file `project-name-0.0.0.tgz` contains my module, which has been build with `ngc` and includes the generated metadata.

[without ngc](https://github.com/m-abs/angular2-aot-demo/tree/without-ngc)
`npm run ngc` on the demo project fails with the error above.
The only difference between the two, is that the code in `project-name-0.0.0.tgz` have only been built with `tsc` and doesn't include the metadata generated by `ngc`.

> BTW, in your setup, you don't benefit from Tree Shaking, since you're transpiling using tsc.

I thought that would be done by `webpack` in the app which import my module (e.g. the demo app's webpack)?

> Anyways, building and distributing an ng2 NgModule is no different from sharing a typescript package.

Normally I would just transpile my `*.ts`-files into `*.js` + `*.d.ts`-files and distribute them to `npmjs`, but when I do that and imports it into the demo project, `npm run ngc` fails.
